codex/sales/reservations.nim

```diff
@@ -71,7 +74,8 @@ type
     size* {.serialize.}: UInt256
     requestId* {.serialize.}: RequestId
     slotIndex* {.serialize.}: UInt256
-  Reservations* = ref object
+  Reservations* = ref object of RootObj
+    locks: Table[AvailabilityId, AsyncLock]
```

Comment:
I'm not a big fan of this construction; this table will keep growing and elements are never removed. Wouldn't it make more sense to add a `lock` field to the `Availability` type?

Reply:
Yeah you are right that it will never be removed, that should be fixed, but I am not sure how the `lock` field would work here, because `Availability` is created from the JSON datastore backend... Every time you call `reservations.get()` you will get new object and unless we would persist the lock, which IMHO is bad idea, then you would get unsynced locks across the Availability's objects. Sooo... instead I would suggest removing the entry upon deleting availability, which BTW we still do not support 🤪 

Btw. this table will be very small as there will be generally very low number of Availabilities ;-)

Reply:
Yes, we should definitely not persist the lock :grinning: But I think that the need to introduce this table indicates that the architecture of Reservations is not working for this use case. The table feels like a workaround, and instead of making the workaround better (by deleting entries etc.) I'd rather we take a good look at the architecture and fix it.

The nim garbage collector is very good at removing things that are no longer needed, so that's why I was thinking of adding the lock to a garbage collected object. But as you pointed out, you get a fresh object (and therefore a fresh lock) every time you try to perform an operation on an availability. And I currently don't see a good way to have unique objects in memory for each availability.

So I'd like to suggest the following: remove the table of locks and replace it with a single lock. This makes it very simple to reason about. The only downside is that you cannot update two availabilities in parallel, but that is ok because they'll simply be processed one after the other.

Reply:
> So I'd like to suggest the following: remove the table of locks and replace it with a single lock. This makes it very simple to reason about. The only downside is that you cannot update two availabilities in parallel, but that is ok because they'll simply be processed one after the other.

When I started reading your first comment, I thought you would suggest this, so I already thought about it.

Honestly, I don't see why you consider this a bad workaround. From my POV this seems like a legitimate solution, that is trying to create the least obstruction to the concurrency that we have thanks to Eric's Slot Queue. And having a single lock for all the availabilities might hinder this quite a bit. I would understand your concern if we would be talking about a Table with potentially thousands of records, but realistically we will have tens of Availabilities and hence this Table will be very small.